File: golang/matrix.go
package main

// Matrix Operations to ease kalman filter.

import (
	"errors"
)

var ErrDimensions = errors.New("columns of a and rows of b should be equal")

// MatMul: Multiplies to Matrices
func MatMul(A, B [][]float64) [][]float64 {

	rowA, colA, rowB, colB := len(A), len(A[0]), len(B), len(B[0])

	if colA != rowB {
		panic("not equal")
	}

	result := make([][]float64, rowA)
	for i := range result {
		result[i] = make([]float64, colB)
	}

	for row := range result {
		for col := range result[row] {
			for i := range colA {
				result[row][col] += A[row][i] * B[i][col]
			}
		}
	}

	return result
}

// MatT: Transpose Matrix
func MatT(A [][]float64) [][]float64 {
	rowA, colA := len(A), len(A[0])

	result := make([][]float64, colA)
	for i := range colA {
		result[i] = make([]float64, rowA)
	}

	for row := range rowA {
		for col := range colA {
			result[col][row] = A[row][col]
		}
	}

	return result
}

// MatDiv: Divides two matrices
func MatDiv(A, B [][]float64) [][]float64 {
	return MatMul(A, MatT(B))
}

// MatAdd: Adds two matrices
func MatAdd(A, B [][]float64) [][]float64 {
	if len(A) != len(B) || len(A[0]) != len(B[0]) {
		panic("dimensions error")
	}

	result := clone(A)
	for i := range len(result) {
		for j := range len(result[0]) {
			result[i][j] += B[i][j]
		}
	}

	return result
}

// MatSub: Subtract two matrices
func MatSub(A, B [][]float64) [][]float64 {
	if len(A) != len(B) || len(A[0]) != len(B[0]) {
		panic("dimensions error")
	}

	result := clone(A)
	for i := range len(result) {
		for j := range len(result[0]) {
			result[i][j] -= B[i][j]
		}
	}

	return result
}

// MatMulC: Multiples a matrix with a constant
func MatMulC(A [][]float64, C float64) [][]float64 {
	result := clone(A)

	for i := range len(result) {
		for j := range len(result[0]) {
			result[i][j] *= C
		}
	}
	return result
}

// MatI: Produce Identity Matrix
func MatI(Dimension int) [][]float64 {
	I := make([][]float64, Dimension)

	for i := range Dimension {
		I[i] = make([]float64, Dimension)
	}

	for i := range len(I) {
		for j := range len(I) {
			if i == j {
				I[i][j] = 1
			}
		}
	}

	return I
}

// MatDiffuse: Converts a 1 * 1 matrix to a float
// Same working as A[0][0], used for abstraction purposes.
func MatDiffuse(A [][]float64) float64 {
	return A[0][0]
}

// clone: makes a new copy of the matrix.
// this prevents editing the copied matrix in the operations.
func clone(A [][]float64) [][]float64 {
	result := make([][]float64, len(A))
	for i := range result {
		result[i] = make([]float64, len(A[0]))
	}

	for i := range A {
		for j := range A[i] {
			result[i][j] = A[i][j]
		}
	}

	return result
}

File: c/matrix.c
#include "kalman.h"


// MatMul: Multiplies two matrices
float** MatMul(float** A, float** B, int rowA, int colA, int rowB, int colB) {
    if (colA != rowB) {
        return NULL;
    }

    float** result = (float**)malloc(rowA * sizeof(float*));
    for (int i = 0; i < rowA; i++) {
        result[i] = (float*)calloc(colB, sizeof(float));
    }

    for (int row = 0; row < rowA; row++) {
        for (int col = 0; col < colB; col++) {
            for (int i = 0; i < colA; i++) {
                result[row][col] += A[row][i] * B[i][col];
            }
        }
    }

    return result;
}

// MatT: Transpose Matrix
float** MatT(float** A, int rowA, int colA) {
    float** result = (float**)malloc(colA * sizeof(float*));
    for (int i = 0; i < colA; i++) {
        result[i] = (float*)malloc(rowA * sizeof(float));
    }

    for (int row = 0; row < rowA; row++) {
        for (int col = 0; col < colA; col++) {
            result[col][row] = A[row][col];
        }
    }

    return result;
}

// MatDiv: Divides two matrices (Multiplication with transpose)
float** MatDiv(float** A, float** B, int rowA, int colA, int rowB, int colB) {
    float** Bt = MatT(B, rowB, colB);
    float** result = MatMul(A, Bt, rowA, colA, colB, rowB);
    for (int i = 0; i < colB; i++) {
        free(Bt[i]);
    }
    free(Bt);
    return result;
}

// MatAdd: Adds two matrices
float** MatAdd(float** A, float** B, int rowA, int colA) {
    float** result = (float**)malloc(rowA * sizeof(float*));
    for (int i = 0; i < rowA; i++) {
        result[i] = (float*)malloc(colA * sizeof(float));
    }

    for (int i = 0; i < rowA; i++) {
        for (int j = 0; j < colA; j++) {
            result[i][j] = A[i][j] + B[i][j];
        }
    }

    return result;
}

// MatSub: Subtracts two matrices
float** MatSub(float** A, float** B, int rowA, int colA) {
    float** result = (float**)malloc(rowA * sizeof(float*));
    for (int i = 0; i < rowA; i++) {
        result[i] = (float*)malloc(colA * sizeof(float));
    }

    for (int i = 0; i < rowA; i++) {
        for (int j = 0; j < colA; j++) {
            result[i][j] = A[i][j] - B[i][j];
        }
    }

    return result;
}

// MatMulC: Multiplies a matrix with a constant
float** MatMulC(float** A, int rowA, int colA, float C) {
    float** result = (float**)malloc(rowA * sizeof(float*));
    for (int i = 0; i < rowA; i++) {
        result[i] = (float*)malloc(colA * sizeof(float));
    }

    for (int i = 0; i < rowA; i++) {
        for (int j = 0; j < colA; j++) {
            result[i][j] = A[i][j] * C;
        }
    }

    return result;
}

// MatI: Produces Identity Matrix
float** MatI(int Dimension) {
    float** I = (float**)malloc(Dimension * sizeof(float*));
    for (int i = 0; i < Dimension; i++) {
        I[i] = (float*)calloc(Dimension, sizeof(float));
    }

    for (int i = 0; i < Dimension; i++) {
        I[i][i] = 1;
    }

    return I;
}

// MatDiffuse: Converts a 1 * 1 matrix to a float
float MatDiffuse(float** A) {
    return A[0][0];
}


// Helper function to free matrix memory
void freeMatrix(float** matrix, int rows) {
    for (int i = 0; i < rows; i++) {
        free(matrix[i]);
    }
    free(matrix);
}

// Helper function to print matrix
void printMatrix(float** matrix, int rows, int cols) {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            printf("%f ", matrix[i][j]);
        }
        printf("\n");
    }
}


// Example usage
int main() {
    float* Adata[] = { (float[]){1, 2}, (float[]){3, 4} };
    float** A = (float**)Adata;

    float* Bdata[] = { (float[]){5, 6}, (float[]){7, 8} };
    float** B = (float**)Bdata;

    int rowA = 2, colA = 2, rowB = 2, colB = 2;

    float** result = MatMul(A, B, rowA, colA, rowB, colB);
    printMatrix(result, rowA, colB);
    freeMatrix(result, rowA);

    return 0;
}

